File: JS/LeetC/arrays/heightChecker.js
// O(n) solution - use ordered histogram technique

// Since we know that heights will be in [1, 100] range, we can generate an ordered histogram (using an array) of the count of each number (value at arr[i] is the count of i in original array).
// now when we loop through the heights array, we check if each element has already been encountered based on its remainder count.
function heightChecker(arr) {
	let map = Array.from({ length: 101 }, (_) => 0);

	for (let e of arr) {
		map[e]++;
	}

	let count = 0;

	let hPtr = 1;

	for (let h of arr) {
		while (map[hPtr] === 0) {
			hPtr++;
		}

		if (h !== hPtr) {
			count++;
		}

		map[hPtr]--;
	}

	return count;
}


// O(nlogn) solution - sort and compare

// function heightChecker(arr) {
// 	let copy = arr.slice().sort((a, b) => a - b);

// 	let count = 0;

// 	for (let i = 0; i < arr.length; i++) {
// 		if (arr[i] !== copy[i]) {
// 			count++;
// 		}
// 	}

// 	return count;
// }


// console.log(heightChecker([1, 1, 4, 2, 1, 3]));
// console.log(heightChecker([5,1,2,3,4]));
console.log(heightChecker([1,2,3,4,5]));

File: python/leetcode/arrays/heightChecker.py
class Solution:

  # def heightChecker(self, heights):

  #   sorted = list(heights)

  #   sorted.sort()

  #   count = 0

  #   for i in range(len(heights)):

  #     if not sorted[i] == heights[i]:
  #       count += 1

  #   return count

  def heightChecker(self, heights):
    freq = [0] * 101

    for n in heights:
      freq[n] += 1

    count = 0
    hPtr = 1

    for h in heights:
      while freq[hPtr] == 0:
        hPtr += 1

      if not h == hPtr:
        count += 1

      freq[hPtr] -= 1

    return count


s = Solution()

s.heightChecker([1,1,4,2,1,3])